File: Quadtree/QuadTree.lua
local AABB = require("AABB2D")

local M_M = {
	__call = function(class, ...)
		return class.new(...)
	end
}
local O_M = {}
local O_I = {}
local M = {}
	M.MAX_ITEMS = 4
	M.MAX_DEPTH = 20
	M.new = function(aabb, depth)
		local self = setmetatable({}, O_M)
		self.items = {}
		self.aabb = aabb
		self.depth = depth or 0
		--self.NW
		--self.NE
		--self.SW
		--self.SE
		return self
	end
O_I.split = function(self)
	if not (self.depth < M.MAX_DEPTH) then return false end
	local hw = self.aabb.w/2
	local hh = self.aabb.h/2

	self.NW = M(AABB(self.aabb.x,				self.aabb.y,			hw, hh), depth+1)
	self.NE = M(AABB(self.aabb.x + hw,	self.aabb.y,			hw, hh), depth+1)
	self.SW = M(AABB(self.aabb.x,				self.aabb.y + hh, hw, hh), depth+1)
	self.SE = M(AABB(self.aabb.x + hw, 	self.aabb.y + hh, hw, hh), depth+1)

	return true
end
O_I.insert = function(self, item)
	if not self.aabb:contains(item.point) then return false end
	if #self.items < M.MAX_ITEMS then
		table.insert(self.items, item)
		return true
	end

	if not self.NE then
		if not self:split() then return false end
	end

	if self.NE:insert(item) then return true end
	if self.NW:insert(item) then return true end
	if self.SW:insert(item) then return true end
	if self.SE:insert(item) then return true end

	return false
end
O_I.range_query = function(self, range)
	if not range:intersects(self.aabb) then return false end
	local found = {}

	for _,item in ipairs(self.items) do
		if range:contains(item.point) then
			table.insert(found, item)
		end
	end

	if not self.NW then return found end

	for _,item in ipairs(self.NW:range_query(range)) do table.insert(found, item) end
	for _,item in ipairs(self.NE:range_query(range)) do table.insert(found, item) end
	for _,item in ipairs(self.SW:range_query(range)) do table.insert(found, item) end
	for _,item in ipairs(self.SE:range_query(range)) do table.insert(found, item) end

	return items
end

O_M.__index = O_I
setmetatable(M, M_M)

return M

File: Quadtree/Quadtree.py
#!/usr/bin/python3

import random

class Point:
	def __init__(self, x, y):
		self.x = x
		self.y = y
	def __str__(self):
		return "(" + str(self.x) + "," + str(self.y) + ")"
	def __repr__(self):
		return "(" + str(self.x) + ", " + str(self.y) + ")"

class Bound:
	def __init__(self, x, y, w, h):
		self.x = x
		self.y = y
		self.w = w
		self.h = h
	def contains(self, point):
		return ((point.x > self.x) and
						(point.y > self.y) and
						(point.x < self.x + self.w) and
						(point.y < self.y + self.h))
	def intersects(self, bound):
		return ((self.x < bound.x + bound.w) and
						(self.y < bound.y + bound.h) and
						(bound.x < self.x + self.w) and
						(bound.y < self.y + self.h))

class Quadtree:
	MAXDEPTH = 20
	MAXITEMS = 4
	NE = 0
	NW = 1
	SW = 2
	SE = 3
	def __init__(self, bound, depth=0):
		self.bound		= bound
		self.depth		= depth
		self.items		= []
		self.children = None
	def insert(self, point, value):
		if not self.bound.contains(point): return False

		if len(self.items) < Quadtree.MAXITEMS:
			self.items.append({'point':point, 'value':value})
			return True

		if not self.children:
			self.split()

		for c in self.children:
			if c.insert(point, value):
				return True
		return False
	def rangequery(self, bound):
		cursor = self
		stack = []
		found = []
		stack.append(self)
		
		while len(stack):
			cursor = stack.pop()
			for i in cursor.items:
				if bound.contains(i['point']):
					found.append(i['point'])
				if cursor.children:
					for c in cursor.children:
						stack.append(c)
		return found
	def split(self):
		if self.depth == Quadtree.MAXDEPTH: return False

		c = [0,1,2,3]
		b = self.bound
		hw, hh = b.w/2,  b.h/2

		c[0] = Quadtree(Bound(b.x + hw, b.y,			hw, hh), self.depth + 1)
		c[1] = Quadtree(Bound(b.x,			b.y,			hw, hh), self.depth + 1)
		c[2] = Quadtree(Bound(b.x, 			b.y + hh, hw, hh), self.depth + 1)
		c[3] = Quadtree(Bound(b.x + hw, b.y + hh,	hw, hh), self.depth + 1)

		self.children = c
		return True




""" Testing """
test = Quadtree(Bound(0,0,128,256))
a = []

for i in range(20):
	a.append(Point(random.randint(0,128), random.randint(0,256)))
	test.insert(a[i], "test")

print(a)
print()
print(test.rangequery(Bound(0,0,64,64)))








